### PR TITLE
First scratch of rexml implementation

### PR DESCRIPTION
**Pull request by fanantoxa**
Related to task: [Update XmlBuilder to use REXML class](https://github.com/soulcutter/saxerator/issues/43)
- [ ] Check that works good
- [ ] Check strange parsing. `&#x2014;` on `xml_builder_spec.rb` was printed as `—` (long dash)
- [ ] Refactor code to be more well designed

@soulcutter Please take a look here. 
That is the first scratch but tests works good. Also I've turned on xml `output_type` for Ox too.

### CODE REVIEW

**Comment by fanantoxa** [comment]
@soulcutter I've update a bit the code. Please check it)

File: lib/saxerator/builder/xml_builder.rb [hunk 1]
```diff
@@ -1,5 +1,11 @@
+require 'rexml/document'
+
 module Saxerator
   module Builder
+    class XmlDocument < REXML::Document
```
Comment (soulcutter) [inline lib/saxerator/builder/xml_builder.rb]: Why is there a need for a `to_xml` alias? Less might be more here - meaning we just return a bare REXML::Document and let end-users transform it however they prefer.

File: lib/saxerator/configuration.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@ class Configuration
 
     ALLOWED_OUTPUT_TYPES = {
       nokogiri: [:hash, :xml],
-      ox: [:hash]
+      ox: [:hash, :xml]
```
Comment (soulcutter) [inline lib/saxerator/configuration.rb]: maybe we don't need constant this anymore?

File: lib/saxerator/builder/xml_builder.rb (hunk 1 above)
Comment (fanantoxa) [inline lib/saxerator/builder/xml_builder.rb]: @soulcutter Before this we had a Nokogiri document and he has a :to_xml method.
I thought that it'll be good to backward compatibility.
If it's bad Idea I can remove it

File: lib/saxerator/configuration.rb (hunk 2 above)
Comment (fanantoxa) [inline lib/saxerator/configuration.rb]: @soulcutter I think we can just check that we have right output_type. I'll  update it.

**Comment by fanantoxa** [comment]
@soulcutter Also question about output. REXML convert `&#x2014;` to  `—` (long dash). Is it ok or I have to check how to save `&#x2014;` on output string?

**Comment by fanantoxa** [comment]
@soulcutter I've removed out_put type restrictions for different parsers.

File: lib/saxerator/builder/xml_builder.rb (hunk 1 above)
Comment (soulcutter) [inline lib/saxerator/builder/xml_builder.rb]: I'm for removing it and simply returning the REXML::Document object. Technically this library isn't 1.0, so this is a tolerable breaking change.

File: README.md [hunk 3]
```diff
@@ -100,7 +100,7 @@ end
 | Setting           | Default | Values          | Description
 |:------------------|:--------|-----------------|------------
 | `adapter`         | `:nokogiri` | `:nokogiri`, `:ox` | The XML parser used by Saxerator |
-| `output_type`     | `:hash` | `:hash`, `:xml` | The type of object generated by Saxerator's parsing. `:hash` should be self-explanatory, `:xml` generates a `Nokogiri::XML::Document`
+| `output_type`     | `:hash` | `:hash`, `:xml` | The type of object generated by Saxerator's parsing. `:hash` should be self-explanatory, `:xml` generates a `REXML::Document` (but have additional method :to_xml)
```
Comment (soulcutter) [inline README.md]: Removing our REXML::Document wrapper means we can remove the documentation of `to_xml` here

File: spec/lib/builder/xml_builder_spec.rb [hunk 4]
```diff
@@ -10,29 +10,9 @@
...
-  <contributor type="primary">
-    <name>Jane Doe</name>
-  </contributor>
-  <contributor>
-    <name>Leviticus Alabaster</name>
-  </contributor>
-</entry>
-    eos
+    expected_xml = '<?xml version=\'1.0\'?><entry><id>1</id><published>2012-01-01T16:17:00-06:00</published><updated>2012-01-01T16:17:00-06:00</updated><link href="https://example.com/blog/how-to-eat-an-airplane"/><title>How to eat an airplane</title><content type="html">&lt;p&gt;Airplanes are very large — this can present difficulty in digestion.&lt;/p&gt;</content><media:thumbnail url="http://www.gravatar.com/avatar/a9eb6ba22e482b71b266daadf9c9a080?s=80"/><author><name>Soul&lt;utter</name></author><contributor type="primary"><name>Jane Doe</name></contributor><contributor><name>Leviticus Alabaster</name></contributor></entry>'
     expect(entry.to_xml).to eq(expected_xml)
```
Comment (soulcutter) [inline spec/lib/builder/xml_builder_spec.rb]: `to_xml` will need to be `to_s` here once the REXML::Document wrapper is removed.

File: spec/lib/saxerator_spec.rb [hunk 5]
```diff
@@ -52,9 +52,9 @@
         specify { expect(parser.all).to eq('bar' => 'baz') }
       end
 
-      context 'with config.output_type = :xml', :nokogiri_only do
+      context 'with config.output_type = :xml' do
         let(:output_type) { :xml }
-        specify { expect(parser.all).to be_a Nokogiri::XML::Document }
+        specify { expect(parser.all).to be_a Saxerator::Builder::XmlDocument }
```
Comment (soulcutter) [inline spec/lib/saxerator_spec.rb]: It'll be a `REXML::Document` when the wrapper is removed

**Comment by soulcutter** [comment]
> REXML convert &#x2014; to — (long dash). Is it ok or I have to check how to save &#x2014; on output string?

I'm fine with keeping with REXML's behavior for that output.

**Comment by fanantoxa** [comment]
@soulcutter I've updated with your additions, But for some reason test fails on ruby 2.0 with error: 
```
An error occurred while installing nokogiri (1.7.0.1), and Bundler cannot
continue.
Make sure that `gem install nokogiri -v '1.7.0.1'` succeeds before bundling.
The command "eval bundle install --without coverage --binstubs --path=${BUNDLE_PATH:-vendor/bundle}" failed. Retrying, 2 of 3.
```

**Comment by fanantoxa** [comment]
@soulcutter Also look like test doesn't works properly because the not test when we changing adapter actually. It's stubbed.
So when I've tried to use it on app if failed because wasn't able to find adapter to require.

**Comment by soulcutter** [comment]
The latest nokogiri requires ruby 2.1.0+

**Comment by soulcutter** [comment]
> So when I've tried to use it on app if failed because wasn't able to find adapter to require.

It should fail in that scenario. You have to have the adapter on the load path if you want to use it. That could be worth clarifying in the documentation. Adapters are external dependencies, not bundled with this library.

**Pull request merged**